### PR TITLE
Finalize No Results Screen

### PR DESCRIPTION
**Pull request by a-szotyori**
**Note:** Filter for `java` when reviewing, a lot of images were updated and added.

Updated the UI of the No Results Screen. Since analysis can complete also while in the Review Screen the No Results Screen can be shown from there, too.

Updated the example apps to showcase the usage of the No Results Screen.

### How to test
Run the example apps and use images and pdfs that don't contain Pay 5 information to try out the following scenarios:
* Take picture: should show the No Results Screen from Review Screen (wait a little for the analysis to complete) and Analysis Screen
* Import picture: same as above
* Import PDF: should NOT show the No Results Screen (might show a different one in the Screen API Example App)
* Open with image: should show the No Results Screen from Review Screen (wait a little for the analysis to complete) and Analysis Screen WITHOUT the bottom button to go back to the camera
* Open with PDF: same as Import PDF

### Merging
Automatic.

### Ticket 
Issue #86

### CODE REVIEW

File: componentapiexample/src/main/java/net/gini/android/vision/component/GiniVisionActivity.java [hunk 1]
```diff
@@ -351,13 +345,28 @@ public boolean onOptionsItemSelected(MenuItem item) {
...
-        // As the library requests us to go to the Analysis Screen we should only remove the
-        // listener.
-        // We should not cancel the analysis here as we don't know, if we proceed because the
-        // analysis didn't complete or
-        // the user rotated the image
-        getSingleDocumentAnalyzer().removeListener();
-        showFragment(getAnalysisFragment(document));
+        // If we haven't received any valid extractions we query the Gini Vision Library
+        // whether we should show the the Gini Vision No Results Screen
+        if (mNoExtractionsFound && GiniVisionCoordinator.shouldShowGiniVisionNoResultsScreen(document)) {
```
Comment (a-szotyori) [inline componentapiexample/src/main/java/net/gini/android/vision/component/GiniVisionActivity.java]: `onProceedToAnalysisScreen` is called when the user pressed Next in the Review Screen before analysis was reported as finished (with `ReviewFragment#onDocumentAnalyzed()`).

We have to use the `mNoExtractionsFound` bool flag to know if our analysis finished w/o valid results and we should try to show the No Results Screen.

File: componentapiexample/src/main/java/net/gini/android/vision/component/GiniVisionActivity.java [hunk 2]
```diff
@@ -602,22 +581,23 @@ private void showCameraOverlays() {
...
     private void showExtractions(net.gini.android.models.Document giniApiDocument,
-            Map<String, SpecificExtraction> extractions) {
+            Map<String, SpecificExtraction> extractions, Document document) {
         LOG.debug("Show extractions");
-        // We display only the Pay5 extractions: paymentRecipient, iban, bic, amount and
-        // paymentReference
-        if (pay5ExtractionsAvailable(extractions)) {
+        // If we have no Pay 5 extractions we query the Gini Vision Library
+        // whether we should show the the Gini Vision No Results Screen
+        if (hasNoPay5Extractions(extractions.keySet())
```
Comment (a-szotyori) [inline componentapiexample/src/main/java/net/gini/android/vision/component/GiniVisionActivity.java]: Component API: This is where we check the extractions and show the No Results Screen if allowed.

File: componentapiexample/src/main/java/net/gini/android/vision/component/GiniVisionAppCompatActivity.java [hunk 3]
```diff
@@ -596,22 +584,23 @@ private void showCameraOverlays() {
...
     private void showExtractions(net.gini.android.models.Document giniApiDocument,
-            Map<String, SpecificExtraction> extractions) {
+            Map<String, SpecificExtraction> extractions, Document document) {
         LOG.debug("Show extractions");
-        // We display only the Pay5 extractions: paymentRecipient, iban, bic, amount and
-        // paymentReference
-        if (pay5ExtractionsAvailable(extractions)) {
+        // If we have no Pay 5 extractions we query the Gini Vision Library
+        // whether we should show the the Gini Vision No Results Screen
+        if (hasNoPay5Extractions(extractions.keySet())
```
Comment (a-szotyori) [inline componentapiexample/src/main/java/net/gini/android/vision/component/GiniVisionAppCompatActivity.java]: Component API: This is where we check the extractions and show the No Results Screen if allowed.

File: ginivision/src/main/java/net/gini/android/vision/GiniVisionCoordinator.java [hunk 4]
```diff
@@ -24,6 +25,10 @@
 
     private static final Logger LOG = LoggerFactory.getLogger(GiniVisionCoordinator.class);
 
+    public static boolean shouldShowGiniVisionNoResultsScreen(final Document document) {
```
Comment (a-szotyori) [inline ginivision/src/main/java/net/gini/android/vision/GiniVisionCoordinator.java]: Helper method which contains the rules for allowing the showing of our No Results Screen.

File: ginivision/src/main/java/net/gini/android/vision/document/DocumentFactory.java [hunk 5]
```diff
@@ -45,4 +45,9 @@ public static GiniVisionDocument newDocumentFromIntent(@NonNull final Intent int
     public static GiniVisionDocument newDocumentFromPhoto(@NonNull final Photo photo) {
         return ImageDocument.fromPhoto(photo);
     }
+
+    public static GiniVisionDocument newDocumentFromPhotoAndDocument(@NonNull final Photo photo,
```
Comment (a-szotyori) [inline ginivision/src/main/java/net/gini/android/vision/document/DocumentFactory.java]: Had to add a new factory to be able to pass on non-Photo relevant data from an existing Document to the new one. In this case we need to pass on the Intent to know where the Document came from.

File: ginivision/src/main/java/net/gini/android/vision/document/GiniVisionDocument.java [hunk 6]
```diff
@@ -152,7 +152,7 @@ public GiniVisionDocument createFromParcel(Parcel in) {
         }
 
         mType = (Type) in.readSerializable();
-        mIntent = in.readParcelable(Uri.class.getClassLoader());
+        mIntent = in.readParcelable(Intent.class.getClassLoader());
```
Comment (a-szotyori) [inline ginivision/src/main/java/net/gini/android/vision/document/GiniVisionDocument.java]: 🙈

File: ginivision/src/main/java/net/gini/android/vision/noresults/NoResultsFragmentImpl.java [hunk 7]
```diff
@@ -1,50 +1,72 @@
...
+            });
+        } else {
+            backButton.setVisibility(GONE);
+        }
         return view;
     }
 
+    private boolean isDocumentFromCameraScreen() {
+        final Intent intent = mDocument.getIntent();
+        return intent == null || intent.getAction() == null;
```
Comment (a-szotyori) [inline ginivision/src/main/java/net/gini/android/vision/noresults/NoResultsFragmentImpl.java]: Only Intents received with "Open With" have an action.

File: ginivision/src/main/java/net/gini/android/vision/review/ReviewActivity.java [hunk 8]
```diff
@@ -247,11 +272,25 @@ private void showFragment() {
 
     @Override
     public void onProceedToAnalysisScreen(@NonNull Document document) {
-        mAnalyzeDocumentActivityIntent.putExtra(AnalysisActivity.EXTRA_IN_DOCUMENT, document);
-        if (mDocumentAnalysisErrorMessage != null) {
-            mAnalyzeDocumentActivityIntent.putExtra(AnalysisActivity.EXTRA_IN_DOCUMENT_ANALYSIS_ERROR_MESSAGE, mDocumentAnalysisErrorMessage);
+        if (mNoExtractionsFound) {
```
Comment (a-szotyori) [inline ginivision/src/main/java/net/gini/android/vision/review/ReviewActivity.java]: If we were told, that no extractions were found we can check whether we should show the No Results Screen. If we shouldn't we simply finish and return to the Camera Screen. If the No Results Screen is shown we also finish the Review Screen so that when the No Results Screen is closed we end up back at the Camera Screen.

File: componentapiexample/src/main/java/net/gini/android/vision/component/GiniVisionActivity.java [hunk 9]
```diff
@@ -126,11 +126,10 @@ public void onExtractionsReceived(Map<String, SpecificExtraction> extractions) {
                             analysisFragment.onDocumentAnalyzed();
                             stopScanAnimation();
                             showExtractions(getSingleDocumentAnalyzer().getGiniApiDocument(),
-                                    extractions);
+                                    extractions, document);
                         } else {
-                            LOG.debug(
-                                    "Document analyzed in the Analysis Screen, but not in the "
-                                            + "Analysis Screen anymore.");
+                            LOG.debug("Document analyzed in the Analysis Screen, but not in the "
```
Comment (kikettas) [inline componentapiexample/src/main/java/net/gini/android/vision/component/GiniVisionActivity.java]: This log message does not make sense. I think that some parameter is missing, otherwise both strings should be joined.

File: componentapiexample/src/main/java/net/gini/android/vision/component/GiniVisionActivity.java [hunk 10]
```diff
@@ -357,7 +351,12 @@ public void onProceedToAnalysisScreen(@NonNull Document document) {
         // analysis didn't complete or
         // the user rotated the image
         getSingleDocumentAnalyzer().removeListener();
-        showFragment(getAnalysisFragment(document));
+        pushFragment(getAnalysisFragment(document));
+    }
+
+    private void showNoResultsScreen(final @NonNull Document document) {
+        getFragmentManager().popBackStack(getCameraFragment().getClass().getSimpleName(), 0);
```
Comment (kikettas) [inline componentapiexample/src/main/java/net/gini/android/vision/component/GiniVisionActivity.java]: Just a question here. I do not remember well about the popBackStack method but this is done asynchronously, right? So, will be there any issue if next fragment is pushed just after this line?
I see that after that inside `pushFragment`, the new `fragment`is added, so in case that there would be any issue concat both or use `popBackStackImmediate()` instead.

File: componentapiexample/src/main/java/net/gini/android/vision/component/GiniVisionActivity.java [hunk 11]
```diff
@@ -401,18 +400,17 @@ public void onExtractionsReceived(Map<String, SpecificExtraction> extractions) {
...
                             mExtractionsFromReviewScreen = extractions;
+                            // Calling onDocumentAnalyzed() is important to notify the Review
+                            // Fragment that the
+                            // analysis has completed successfully
+                            reviewFragment.onDocumentAnalyzed();
                         } else {
-                            LOG.debug(
-                                    "Document analyzed in the Review Screen, but not in the "
-                                            + "Review Screen anymore.");
+                            LOG.debug("Document analyzed in the Review Screen, but not in the "
```
Comment (kikettas) [inline componentapiexample/src/main/java/net/gini/android/vision/component/GiniVisionActivity.java]: Same as above.

File: componentapiexample/src/main/java/net/gini/android/vision/component/GiniVisionActivity.java [hunk 12]
```diff
@@ -429,29 +427,22 @@ protected void onStart() {
...
         LOG.debug("Remove the Onboarding Screen");
         showCameraOverlays();
         Fragment fragment = getFragmentManager().findFragmentById(
                 R.id.fragment_container_onboarding);
         if (fragment != null) {
-            getFragmentManager().beginTransaction()
-                    .setCustomAnimations(R.animator.fade_in, R.animator.fade_out)
-                    .remove(fragment)
-                    .commit();
+            getFragmentManager().beginTransaction().setCustomAnimations(R.animator.fade_in,
```
Comment (kikettas) [inline componentapiexample/src/main/java/net/gini/android/vision/component/GiniVisionActivity.java]: Why did you change the indentation? For me it is more readable in the previous way.

File: ginivision/src/main/java/net/gini/android/vision/noresults/NoResultsActivity.java [hunk 13]
```diff
@@ -53,22 +61,42 @@ public void onBackToCameraPressed() {
...
+        if (extras != null) {
+            mDocument = extras.getParcelable(EXTRA_IN_DOCUMENT);
+        }
+        checkRequiredExtras();
+    }
+
+    private void checkRequiredExtras() {
+        if (mDocument == null) {
+            throw new IllegalStateException(
+                    "NoResultsActivity requires a Document. Set it as an extra using the EXTRA_IN_DOCUMENT key.");
```
Comment (kikettas) [inline ginivision/src/main/java/net/gini/android/vision/noresults/NoResultsActivity.java]: Why this exception is thrown but not caught afterwards? Did not Android Studio complain when `checkRequiredExtras` was called?

File: ginivision/src/main/java/net/gini/android/vision/noresults/NoResultsFragmentHelper.java [hunk 14]
```diff
@@ -0,0 +1,36 @@
...
+        arguments.putParcelable(ARGS_DOCUMENT, document);
+        return arguments;
+    }
+
+    static NoResultsFragmentImpl createFragmentImpl(@NonNull FragmentImplCallback fragment, @NonNull Bundle arguments) {
+        Document document = arguments.getParcelable(ARGS_DOCUMENT);
+        if (document != null) {
+            return new NoResultsFragmentImpl(fragment, document);
+        } else {
+            throw new IllegalStateException("NoResultsFragmentCompat and NoResultsFragmentStandard require a Document. Use the createInstance() method of these classes for instantiating.");
```
Comment (kikettas) [inline ginivision/src/main/java/net/gini/android/vision/noresults/NoResultsFragmentHelper.java]: Here as well. Maybe it is me that I'm wrong about try-catch on Java, but I would swear that Android Studio complained about it.

File: ginivision/src/main/java/net/gini/android/vision/noresults/NoResultsFragmentImpl.java [hunk 15]
```diff
@@ -1,50 +1,72 @@
...
-        });
+        if (isDocumentFromCameraScreen()) {
+            backButton.setOnClickListener(new View.OnClickListener() {
+                @Override
+                public void onClick(final View view) {
+                    mListener.onBackToCameraPressed();
+                }
+            });
+        } else {
+            backButton.setVisibility(GONE);
```
Comment (kikettas) [inline ginivision/src/main/java/net/gini/android/vision/noresults/NoResultsFragmentImpl.java]: Maybe I have misunderstood something about we discussed before, but the back button should be replaced with a close button, right?

File: ginivision/src/main/java/net/gini/android/vision/review/ReviewActivity.java [hunk 16]
```diff
@@ -291,6 +330,11 @@ public void onDocumentAnalyzed() {
     }
 
     @Override
+    public void noExtractionsFound() {
```
Comment (kikettas) [inline ginivision/src/main/java/net/gini/android/vision/review/ReviewActivity.java]: The method name is a little bit confusing for me, shouldn't be something like `setNoExtractionsFound(Bool)` or `onNoExtractionsFound` (just suggestions)?

File: ginivision/src/main/java/net/gini/android/vision/review/ReviewFragmentImpl.java [hunk 17]
```diff
@@ -113,6 +113,10 @@ public void onDocumentAnalyzed() {
         mDocumentWasAnalyzed = true;
     }
 
+    @Override
+    public void noExtractionsFound() {
```
Comment (kikettas) [inline ginivision/src/main/java/net/gini/android/vision/review/ReviewFragmentImpl.java]: Same as above.

File: ginivision/src/main/java/net/gini/android/vision/review/ReviewFragmentImpl.java [hunk 18]
```diff
@@ -425,27 +433,24 @@ public void onFailed() {
...
+        mListener.onProceedToAnalysisScreen(
+                DocumentFactory.newDocumentFromPhotoAndDocument(mPhoto, mDocument));
     }
 
     private void applyRotationToPhoto(@NonNull PhotoEdit.PhotoEditCallback callback) {
         if (mPhoto == null) {
             return;
         }
         LOG.debug("Rotating the Photo {} degrees", mCurrentRotation);
-        mPhoto.edit()
```
Comment (kikettas) [inline ginivision/src/main/java/net/gini/android/vision/review/ReviewFragmentImpl.java]: As mentioned above, for me it is more readable this that having all in one line.

File: ginivision/src/main/res/layout-sw600dp-land/gv_fragment_noresults.xml [hunk 19]
```diff
@@ -1,136 +1,217 @@
...
 
-    <LinearLayout
+    <TextView
+        android:id="@+id/gv_no_results_header"
+        style="@style/GiniVisionTheme.NoResults.Header.TextStyle"
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:layout_alignLeft="@+id/gv_no_results_headline"
+        android:layout_alignParentTop="true"
+        android:layout_alignStart="@+id/gv_no_results_headline"
```
Comment (kikettas) [inline ginivision/src/main/res/layout-sw600dp-land/gv_fragment_noresults.xml]: I'm not sure but I can remember that Android Studio was always "suggesting" to use `alignStart` instead of `alignLeft`. Was it because I was using a minimum SDK higher than the minimum required to use only `alignStart`?

File: ginivision/src/main/res/layout-sw600dp-land/gv_fragment_noresults.xml [hunk 20]
```diff
@@ -1,136 +1,217 @@
...
+                <TextView
+                    style="@style/GiniVisionTheme.NoResults.Tip.TextStyle"
+                    android:layout_width="wrap_content"
+                    android:layout_height="match_parent"
+                    android:layout_gravity="center_horizontal"
+                    android:layout_marginBottom="@dimen/gv_noresults_tip_text_bottom_margin"
+                    android:layout_marginEnd="@dimen/gv_noresults_tip_text_right_margin"
+                    android:layout_marginLeft="@dimen/gv_noresults_tip_text_left_margin"
+                    android:layout_marginRight="@dimen/gv_noresults_tip_text_right_margin"
+                    android:layout_marginStart="@dimen/gv_noresults_tip_text_left_margin"
```
Comment (kikettas) [inline ginivision/src/main/res/layout-sw600dp-land/gv_fragment_noresults.xml]: Seeing this I can guess your answer in my previous question...

File: componentapiexample/src/main/java/net/gini/android/vision/component/GiniVisionActivity.java (hunk 10 above)
Comment (a-szotyori) [reply componentapiexample/src/main/java/net/gini/android/vision/component/GiniVisionActivity.java]: The FragmentManager works with transactions so popping and showing the new fragment will be scheduled and executed in order.

File: componentapiexample/src/main/java/net/gini/android/vision/component/GiniVisionActivity.java (hunk 12 above)
Comment (a-szotyori) [reply componentapiexample/src/main/java/net/gini/android/vision/component/GiniVisionActivity.java]: It wasn't me (:P), the code style made this. I fixed it now to preserve line breaks.

File: ginivision/src/main/java/net/gini/android/vision/noresults/NoResultsActivity.java (hunk 13 above)
Comment (a-szotyori) [reply ginivision/src/main/java/net/gini/android/vision/noresults/NoResultsActivity.java]: That's a RuntimeException and doesn't have to be catched. Shouldn't happen in production or if it does there is a bigger issue of why is there no document anymore. If clients use this Activity and don't pass in the Document they will see this exception.

File: ginivision/src/main/java/net/gini/android/vision/noresults/NoResultsFragmentHelper.java (hunk 14 above)
Comment (a-szotyori) [reply ginivision/src/main/java/net/gini/android/vision/noresults/NoResultsFragmentHelper.java]: The compiler complains about Exceptions, but RuntimeExceptions don't have to be handled. There are likely static code analysers that can warn about unhandled runtime exceptions, but haven't seen it in AS until now.

File: ginivision/src/main/java/net/gini/android/vision/noresults/NoResultsFragmentImpl.java (hunk 15 above)
Comment (a-szotyori) [reply ginivision/src/main/java/net/gini/android/vision/noresults/NoResultsFragmentImpl.java]: 🤔 really? I understood we should simply hide the button. We also don't have a close button design.

File: ginivision/src/main/res/layout-sw600dp-land/gv_fragment_noresults.xml (hunk 19 above)
Comment (a-szotyori) [reply ginivision/src/main/res/layout-sw600dp-land/gv_fragment_noresults.xml]: Yes, for backwards compatibility we have to use both.

File: ginivision/src/main/java/net/gini/android/vision/noresults/NoResultsFragmentHelper.java (hunk 14 above)
Comment (kikettas) [reply ginivision/src/main/java/net/gini/android/vision/noresults/NoResultsFragmentHelper.java]: Ok, I thought that runtime exception should be caught always! 👍

File: ginivision/src/main/java/net/gini/android/vision/noresults/NoResultsFragmentImpl.java (hunk 15 above)
Comment (kikettas) [reply ginivision/src/main/java/net/gini/android/vision/noresults/NoResultsFragmentImpl.java]: But how do they exit the screen then? Because if they come from Open with there is no button on the bottom.
I'm using the same "X" that we use to exit the ScreenAPI. Maybe on Android is not needed since there is a physical button to go back.
